source/slang/slang-ir-specialize-function-call.cpp

```diff
@@ -559,6 +561,11 @@ struct FunctionParameterSpecializationContext
             // the arguments at the new call site, and
             // don't add anything to the specialization key.
             //
+            // We should also add 2 more things such that our specialization
+            // can handle the corner cases that if the oldBase is a nonuniform
+            // resource and also the data type of oldIndex will be handled correctly.
+            // By doing so, just add the oldIndex to the key of call info.
+            ioInfo.key.vals.add(oldIndex);
```

Comment:
I think you just want to add uniformity and the type of the index to the key, instead of the index inst itself. This will be too conservative, causing us to generate too many specializations.

Reply:
Well, that will work as well.
I have one question, for nonuniform index, can we just add the `oldIndex` directly? I don't know how to represent the uniformity, I guess the `kIROp_NonUniformResourceIndex` inst already a good way to represent it.

For others, I can just use the type inst to represent.

Reply:
We can introduce an OpCode calls `kIROp_NonUniformAttr`, and then form a `IRAttributedType(int, NonUniformAttr)` to represent the type of the index inst, and use this as the key.

Reply:
See `IRBuilder::getAttributedType`

Reply:
I make the change that uses `IRAttributedType` as the key now.